R/MPinv.R

```diff
@@ -0,0 +1,30 @@
+# Function taken from `VCA::MPinv`
+#' Moore-Penrose Generalized Inverse of a Matrix 
+#' 
+#' This function is originally implemented in package 'MASS' as function 
+#' \code{ginv}.  
+#' It was adapted to be able to deal with matrices from the 'Matrix' package, 
+#' e.g. sparse matrices. 
+#' 
+#' @param X (object) two-dimensional, for which a Moore-Penrose inverse 
+#' has to be computed 
+#' @param tol (numeric) tolerance value to be used in comparisons 
+#' 
+#' @return (object) A Moore-Penrose inverse of X. 
+#' 
+#' @author Authors of the 'MASS' package.  
+MPinv <- function (X, tol = sqrt(.Machine$double.eps))  {
+  if (length(dim(X)) > 2L) 
+    stop("'X' must be two-dimensional")
+  
+  Xsvd <- svd(X)
```

Comment:
I think that `svd()` converts sparse matrices to dense matrices under the hood? I have just done some minor performance benchmarks, and it looks to me like MPinv is just as fast as `MASS::ginv()`: 

```r
library(VCA)
library(Matrix)
library(microbenchmark)

N <- 100000
X <- factor(sample(1:1000, N, TRUE))
df <- data.frame(X = X)
mf <- model.frame(~X,df)
mm <- sparse.model.matrix(mf)
dim(mm)

XX <- t(mm) %*% mm
dim(XX)

microbenchmark(
  VCA::MPinv(XX), 
  MASS::ginv(as.matrix(XX)), 
  times = 1
)

# Unit: seconds
# expr      min       lq     mean   median       uq      max
# VCA::MPinv(XX) 5.654788 5.654788 5.654788 5.654788 5.654788 5.654788
# MASS::ginv(as.matrix(XX)) 4.829764 4.829764 4.829764 4.829764 4.829764 4.829764

```

Reply:
What do you think of adding armadillo as a dependency? I have added it to the dev branch of `fwildclusterboot` [here](https://github.com/s3alfisc/fwildclusterboot/blob/7a00a08cd96ffb38efade7ccad47c0b2a8160d14/src/pinv.cpp) (but RccpArma is already a dependency).

Reply:
Updated benchmarks running `fwildclusterboot:::pinv`: 

```r
# expr      min       lq     mean   median       uq
# VCA::MPinv(XX) 5.578639 5.578639 5.578639 5.578639 5.578639
# MASS::ginv(as.matrix(XX)) 5.767423 5.767423 5.767423 5.767423 5.767423
# fwildclusterboot:::pinv(as.matrix(XX)) 2.993501 2.993501 2.993501 2.993501 2.993501

```

Reply:
Last benchmark - Eigen's pinv implementation beats Armadillo, by quite a large margin: 

```r
library(VCA)
library(Matrix)
library(microbenchmark)

N <- 100000
X <- factor(sample(1:1000, N, TRUE))
df <- data.frame(X = X)
mf <- model.frame(~X,df)
mm <- sparse.model.matrix(mf)
dim(mm)

XX <- t(mm) %*% mm
dim(XX)

#x <- sparseEigen::spEigen(XX, q = ncol(XX))


microbenchmark(
  Matrix::solve(XX),
  VCA::MPinv(XX), 
  MASS::ginv(as.matrix(XX)), 
  fwildclusterboot:::pinv(as.matrix(XX)), 
  eigenR = EigenR::Eigen_pinverse(as.matrix(XX)), 
  #sparseEigen = sparseEigen::spEigen(XX, q = ncol(XX)),
  #eigen = eigen(XX),
  times = 1
)

# Unit: milliseconds
# expr       min
# Matrix::solve(XX)   58.4333
# VCA::MPinv(XX) 6007.1988
# MASS::ginv(as.matrix(XX)) 4880.4366
# fwildclusterboot:::pinv(as.matrix(XX)) 2981.8584
# eigenR  779.7848
# lq      mean    median        uq       max neval
# 58.4333   58.4333   58.4333   58.4333   58.4333     1
# 6007.1988 6007.1988 6007.1988 6007.1988 6007.1988     1
# 4880.4366 4880.4366 4880.4366 4880.4366 4880.4366     1
# 2981.8584 2981.8584 2981.8584 2981.8584 2981.8584     1
# 779.7848  779.7848  779.7848  779.7848  779.7848     1
```

Reply:
Implementation [here](https://github.com/stla/EigenR/blob/1ffb3091d4bfa6fb0cd5fbfdaa1cc06f7bb6ea69/src/Inverse.cpp#L39).